src/Shared/HttpSys/RequestProcessing/RequestHeaders.cs

```diff
@@ -240,4 +260,36 @@ public IEnumerable<string> GetValues(string key)
         }
         return HeaderParser.Empty;
     }
+
+    private int GetKnownHeadersKeys(Span<string> observedHeaders)
+    {
+        int observedHeadersCount = 0;
+        var keys = HeaderKeys;
+        while (keys.Length > 0)
+        {
+            var header = (HttpSysRequestHeader)keys[0];
+            if (HasKnownHeader(header))
+            {
+                observedHeaders[observedHeadersCount++] = GetHeaderKeyName(header);
+            }
+            keys = keys.Slice(1);
+        }
+        return observedHeadersCount;
+    }
+
+    private int GetKnownHeadersCount()
+    {
+        int observedHeadersCount = 0;
+        var keys = HeaderKeys;
+        while (keys.Length > 0)
```

Comment:
Why not use a for loop? Slice is an unusual pattern when you're consuming data in order.

Reply:
I will revert to for loop. I generally like the idea that I have to index into 0 always.